Add PagerDutyIntegration CRD

grdryn opened the pull request:
This change adds a PagerDutyIntegration CRD & replaces the
ClusterDeployment and SyncSet controllers with a single
PagerDutyIntegration controller. This will allow more than one
PagerDuty service to be created per cluster, so that more than one SRE
team can receive alerts for different concerns (e.g. layered products
on top of OSD).

## Watches

Since there's only one controller, and there's a many-to-many
relationship between PagerDutyIntegrations and ClusterDeployments,
here's how reconciliation will be triggered:

* Create/Update/Delete of a PagerDutyIntegration will trigger a
  reconcile request for that PagerDutyIntegration.
* Create/Update/Delete of a ClusterDeployment will trigger a reconcile
  request for each PagerDutyIntegration that matches that
  ClusterDeployment (using the spec.clusterDeploymentSelector label
  selector).
* Create/Update/Delete of a ConfigMap/Secret/SyncSet owned by a
  ClusterDeployment will trigger a reconcile request for each
  PagerDutyIntegration that matches that ClusterDeployment.

## Migration / upgrade logic

This change also includes temporary migration logic (which can
optionally be removed easily in a subsequent release) to decide based
on an annotation whether the CR should be used for migration of
resources from the pre-PagerDutyIntegration CRD way of working. This
`pd.openshift.io/legacy` annotation will inform the operator that
there may be clusters that were configured with a pagerduty service
already, and that it should update the names of those resources to the
new format. Once all of the clusters have been migrated, it will
remove the annotation.

jewzaam commented:
PR check needs a fix: https://github.com/openshift/pagerduty-operator/pull/97

grdryn commented:
/cc @georgettica

RiRa12621 commented:
/assign @RiRa12621

RiRa12621 commented:
/hold
For time to sync on a plan going forward

georgettica commented on deploy/crds/pagerduty.openshift.io_pagerdutyintegrations_crd.yaml:
```diff
@@ -0,0 +1,136 @@
+          type: object
+        status:
+          description: PagerDutyIntegrationStatus defines the observed state of PagerDutyIntegration
+          type: object
```
seems your CRD is fine here :)

grdryn replied:
It should be, it's generated from the Go types by the operator-sdk :slightly_smiling_face:

It might be important to hightlight that I _did_make a change after generation though, to work around an issue that Kubernetes/OpenShift v3.11 has with the structured schema stuff that these later versions (0.12+ I believe) add to the CRDs.

The change is made with:

```
yq d -i deploy/crds/pagerduty.openshift.io_pagerdutyintegrations_crd.yaml spec.validation.openAPIV3Schema.type
```

georgettica replied:
this is exactly what I wanted to see. maybe add a `//go:generate` that will regenerate it and do this command?

RiRa12621 commented:
/cancel-hold

RiRa12621 commented:
/cancel hold

RiRa12621 commented:
/hold cancel

georgettica commented on README.md:
````diff
@@ -144,7 +137,7 @@ $ oc create namespace fake-cluster-namespace
 ```
 
-If present, set `status.installed` to true.
+If present, set `spec.installed` to true.
````
why was this moved?

grdryn replied:
Good question! The change here in the README that I've made is just to reflect reality. It looks like it was changed deliberately in the code, but the information here in the README was forgotten.

As to why it was moved in the code, I'm not sure. It looks like it was done in #74 for some reason. The title of that pull request is "Status.Installed is deprecated, we should use Spec.Installed", maybe that means something to somebody?

Actually, given that this operator doesn't own the ClusterDeployment CRD, looking in the hive project is where we should be looking. This change seems to have all the details: https://github.com/openshift/hive/pull/526.

grdryn commented on deploy/crds/pagerduty_v1alpha1_pagerdutyintegration_cr.yaml:
```diff
@@ -0,0 +1,18 @@
+apiVersion: pagerduty.openshift.io/v1alpha1
```
Oh yes, this is just an example. I think the operator-sdk put a dummy one here when generating the initial CRD, and I just updated it as I updated the corresponding CRD. I'll move it to a separate directory to make it less confusing (and in case your question is based on everything in `deploy/crds/` being deployed (this is background knowledge that I don't have)).

grdryn commented on go.mod:
```diff
@@ -3,80 +3,48 @@ module github.com/openshift/pagerduty-operator
 
-// Replaces from https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v08x
-// Pinned to kubernetes-1.13.1
+// Replaces from https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v016x
```
Originally I didn't want to upgrade the operator-sdk at all (primarily because my understanding is that this operator still runs on OpenShift 3.11, and there's a bug in that version of k8s that will reject CRDs with some of the structured schema stuff that's generated by the newer versions of the operator-sdk), but I did it to pull in a newer version of the fake k8s client to be able to update some of the pre-existing tests to work with the new logic.

I could probably revert the operator-sdk upgrade & remove those tests as an alternative, would that be preferable?

RiRa12621 replied:
@jewzaam How do you feel about having the operator-sdk upgrade in here?
I'm not strongly opposed but would like extra eyes

jewzaam replied:
If we feel the upgrade will work on 3.11 I'm not opposed, it sounds like it's required for the testing added.  It could be pulled to a separate PR but I'd want it merged before we deploy these changes if it's adding useful tests.

grdryn replied:
@jewzaam @RiRa12621 I've created #100 just now to give you the option to merge that separately if that's what you want to do. :) I won't make any change around it in this PR until you decide whether that's what you want to do or not, then I'll rebase this PR or not.

grdryn commented on pkg/controller/pagerdutyintegration/clusterdeployment_created.go:
```diff
@@ -23,121 +23,161 @@ import (
-		err := r.client.Update(context.TODO(), instance)
+	if utils.HasFinalizer(cd, finalizer) == false {
+		utils.AddFinalizer(cd, finalizer)
+		err := r.client.Update(context.TODO(), cd)
```
When you say another thread, do you mean another thread in this process (concurrent reconciles or a different controller or a different instance of this controller), or in another operator such as hive? My understanding is that there will only be one reconcile at a time in this controller (we don't override the `MaxConcurrentReconciles` option when the controller is created, and the default is 1). Also, even if there are multiple instances of this operator deployed, only one will be active, after [acquiring the lock](https://github.com/openshift/pagerduty-operator/blob/master/cmd/manager/main.go#L100).

The reason I'm bringing this up is that I'm hesitant to change this, as it would require changing other things -- this function now just returns an error value, and relies on the caller to handle that (either continue or return an empty reconcile result & the error, etc). While I _can_ change this here, I'm hesitant to change too much logic that existed before the changes in this pull request, especially when I'm not from the owning team). WDYT?

jewzaam replied:
Agree, any changes to the CR you're watching must result in existing the reconcile.  Another event is triggered for that CR and will be processed by all watching controllers.  You will have race conditions if you continue.

grdryn replied:
I still don't see where a race condition could occur here (did the possibility of it exist before this change?), but I'll update now in any case.

grdryn replied:
Fixed in latest commit :)

grdryn commented on pkg/controller/pagerdutyintegration/mappers.go:
```diff
@@ -0,0 +1,104 @@
+	for _, pdi := range pdiList.Items {
+		selector, err := metav1.LabelSelectorAsSelector(&pdi.Spec.ClusterDeploymentSelector)
+		if err != nil {
+			return []reconcile.Request{}
```
Good point, I'll change it to just a `continue` here (and similar on L89). How does that sound?

grdryn commented on pkg/controller/pagerdutyintegration/pagerdutyintegration_controller.go:
```diff
@@ -0,0 +1,266 @@
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
```
I like this idea & will change this to use it also, thanks! :slightly_smiling_face: 

I wonder if it's a pattern that the kubebuilder folks would accept if you contributed it? That way all new scaffolded controllers from the operator-sdk and others would use it! :smiley:

grdryn commented on README.md:
```diff
@@ -5,14 +5,14 @@ The PagerDuty operator is used to automate integrating Openshift Dedicated clust
-* PagerDuty's reconcile function watches for the `installed` field of the `ClusterDeployment` CRD and waits for the cluster to finish installation. It also sees if `api.openshift.com/noalerts` label is set on the `ClusterDeployment` of the new cluster being provisioned. 
-  * The `api.openshift.com/noalerts` label is used to disable alerts from the provisioned cluster. This label is typically used on test clusters that do not require immediate attention as a result of critical issues or outages. Therefore, PagerDuty does not continue its actions if it finds this label in the new cluster's `ClusterDeployment`.
-* Once the `installed` field becomes true, PagerDuty creates a secret which contains the integration key required to communicate with PagerDuty Web application.
+* The PagerDutyIntegration controller watches for changes to PagerDutyIntegration CRs, and also for changes to appropriately labelled ClusterDeployment CRs (and ConfigMap/Secret/SyncSet resources owned by such a ClusterDeployment).
```
That's debatable :wink: https://www.grammarly.com/blog/labeled-labelled/

Not a worthwhile debate though, so fixed. :slightly_smiling_face:

grdryn commented:
@RiRa12621 @georgettica Thanks a lot for the reviews! :slightly_smiling_face: 

I believe I've addressed all of the comments in one way or another, if you'd like to take a look again.

RiRa12621 commented:
/assign @jewzaam for approval

grdryn commented:
@RiRa12621 @jewzaam Thanks for the sign-off! :D Note that someone/something will need to create a `PagerDutyIntegration` CR for the details that were previously used in the old flow, and that CR should have the `pd.openshift.io/legacy` annotation to lift the existing resources for existing clusters into the new CR-based flow.

RiRa12621 commented:
it's gonna go to int and stage anyways first, so I'd assume we notice all kinds of "problems" like this before it hits prod

grdryn commented:
This has been rebased now, as the changes to upgrade the operator-sdk version have been merged to master in #100

jewzaam commented:
/approve
